### PR TITLE
Output scaling overall merging statistics in the xia2 style.

### PR DESCRIPTION
**Pull request by jbeilstenedmands**
i.e. The end of scaling output looks something like this. Bonus is that a few additional anomalous quality indicators are now output. Only thing not output that is in xia2 output is the Wilson B factor, as this should be calculated after merging and truncating.

```
            ----------Merging statistics by resolution bin----------           

 d_max  d_min   #obs  #uniq   mult.  %comp       <I>  <I/sI>    r_mrg   r_meas    r_pim   cc1/2   cc_ano
 69.31   3.29  22774   3392    6.71  98.63     532.9    53.1    0.038    0.041    0.016   0.999*   0.353*
  3.29   2.61  22323   3319    6.73  97.53     193.6    36.0    0.050    0.054    0.021   0.998*   0.393*
  2.61   2.28  22557   3259    6.92  96.79      99.3    26.6    0.065    0.071    0.027   0.997*   0.324*
  2.28   2.07  22203   3244    6.84  95.98      67.6    20.3    0.079    0.085    0.032   0.996*   0.281*
  2.07   1.93  21523   3169    6.79  95.25      44.0    15.1    0.099    0.108    0.041   0.992*   0.190*
  1.93   1.81  21673   3188    6.80  94.40      25.1    10.4    0.136    0.147    0.056   0.991*   0.186*
  1.81   1.72  21276   3154    6.75  93.98      15.6     7.3    0.180    0.195    0.074   0.986*   0.162*
  1.72   1.65  20912   3106    6.73  93.13      10.4     5.3    0.231    0.251    0.096   0.978*   0.132*
  1.65   1.58  21666   3123    6.94  92.64       8.4     4.4    0.275    0.297    0.112   0.973*   0.053*
  1.58   1.53  20697   3083    6.71  92.03       6.3     3.5    0.326    0.354    0.135   0.955*   0.106*
  1.53   1.48  20174   3067    6.58  91.61       5.0     2.8    0.382    0.415    0.161   0.939*   0.031
  1.48   1.44  20819   3031    6.87  90.80       3.7     2.2    0.473    0.512    0.194   0.915*   0.034
  1.44   1.40  18344   2987    6.14  90.46       3.2     1.8    0.523    0.571    0.227   0.878*   0.001
  1.40   1.37  13881   2480    5.60  72.83       2.7     1.5    0.584    0.645    0.267   0.812*  -0.001
  1.37   1.33   9578   1749    5.48  52.71       2.4     1.3    0.638    0.704    0.292   0.748*  -0.019
  1.33   1.31   6726   1287    5.23  38.73       2.3     1.2    0.654    0.726    0.306   0.735*   0.007
  1.31   1.28   4850   1001    4.85  29.69       2.0     1.0    0.746    0.833    0.362   0.602*  -0.020
  1.28   1.26   2923    700    4.18  21.03       1.8     0.8    0.854    0.974    0.453   0.403*   0.032
  1.26   1.23   1214    407    2.98  12.37       1.5     0.6    0.897    1.070    0.569   0.282*  -0.156
  1.23   1.21    305    191    1.60   5.64       1.6     0.5    0.741    0.981    0.636   0.490*  -0.408
 69.19   1.21 316418  48937    6.47  72.90      69.3    12.8    0.065    0.070    0.027   0.999*   0.329*


               ----------Summary of merging statistics----------               

                                             Overall    Low     High
High resolution limit                           1.21    3.29    1.21
Low resolution limit                           69.19   69.31    1.23
Completeness                                   72.9    98.6     5.6
Multiplicity                                    6.5     6.7     1.6
I/sigma                                        12.8    53.1     0.5
Rmerge(I)                                     0.065   0.038   0.741
Rmerge(I+/-)                                  0.055   0.031   0.695
Rmeas(I)                                      0.070   0.041   0.981
Rmeas(I+/-)                                   0.065   0.037   0.983
Rpim(I)                                       0.027   0.016   0.636
Rpim(I+/-)                                    0.035   0.019   0.695
CC half                                       0.999   0.999   0.490
Anomalous completeness                         71.6    99.1     1.4
Anomalous multiplicity                          3.3     3.5     1.3
Anomalous correlation                         0.329   0.353  -0.408
Anomalous slope                               0.807
dF/F                                          0.070
dI/s(dI)                                      0.959
Total observations                           316418   22774     305
Total unique                                  48937    3392     191

Writing html report to dials.scale.html
Saving the scaled experiments to scaled.expt
Saving the scaled reflections to scaled.refl
See dials.github.io/dials_scale_user_guide.html for more info on scaling options
```

### CODE REVIEW

**Comment by graeme-winter** [comment]
Well, I obviously approve of the suggestion! 🙂

**Comment by jbeilstenedmands** [comment]
One potential issue is that before applying a resolution cutoff, the high resolution bin will often just be noise, which may be confusing or unsightly. Something we could do to get around this is to use the resolutionizer code to suggest a resolution limit, and then report the high resolution bin using that limit? Thoughts @graeme-winter ?

**Comment by graeme-winter** [comment]
Valid concern, would suggest using the resolutionizer code to determine an outer bin then show 

 - overall
 - overall to chosen limit
 - inner
 - outer to chosen limit

unless %USER% has set limit, in which case use that. By a happy coincidence I think this is exactly what xia2.small_molecule does 🤔

**Comment by jbeilstenedmands** [comment]
Example output when resolution limit within measured range:
```
            ----------Merging statistics by resolution bin----------           

 d_max  d_min   #obs  #uniq   mult.  %comp       <I>  <I/sI>    r_mrg   r_meas    r_pim   cc1/2   cc_ano
 72.29   4.85  35119   2123   16.54  90.96      99.1    35.1    0.125    0.129    0.027   0.998*   0.013
  4.85   3.85  35008   2204   15.88  94.03      83.1    30.5    0.147    0.151    0.033   0.996*   0.027
  3.85   3.36  34497   2206   15.64  95.50      41.3    23.8    0.192    0.197    0.044   0.995*  -0.010
  3.36   3.05  34699   2266   15.31  96.26      20.3    16.0    0.314    0.323    0.071   0.989*   0.029
  3.05   2.83  33927   2202   15.41  95.45      13.9    12.3    0.405    0.417    0.094   0.975*   0.002
  2.83   2.67  34209   2268   15.08  97.38      10.7     9.3    0.526    0.541    0.123   0.963*   0.013
  2.67   2.53  35104   2291   15.32  97.61       8.2     7.3    0.627    0.645    0.145   0.942*  -0.018
  2.53   2.42  32881   2239   14.69  96.22       6.4     5.5    0.762    0.787    0.186   0.384*  -0.008
  2.42   2.33  27903   2214   12.60  94.94       5.1     3.9    0.939    0.974    0.252   0.870*   0.007
  2.33   2.25  20854   2161    9.65  92.47       4.2     2.7    0.966    1.014    0.299   0.599*  -0.021
  2.25   2.18  16347   2062    7.93  88.54       3.6     2.0    1.308    1.394    0.464   0.333*   0.014
  2.18   2.12  12857   2020    6.36  85.99       3.0     1.4    1.097    1.189    0.445   0.006   0.004
  2.12   2.06  10391   1968    5.28  84.07       2.8     1.0    1.156    1.298    0.563   0.007  -0.008
  2.06   2.01   8006   1865    4.29  81.09       2.6     0.7    1.330    1.512    0.695  -0.001   0.016
  2.01   1.97   6184   1847    3.35  78.76       2.7     0.6    1.721    2.070    1.108   0.004   0.040
  1.97   1.92   4482   1673    2.68  71.56       0.4     0.5    1.188    1.414    0.747   0.030   0.005
  1.92   1.89   3037   1454    2.09  62.70       3.3     0.4    1.534    1.968    1.209  -0.004   0.250
  1.89   1.85   1692   1059    1.60  45.61       9.1     0.3    1.663    2.162    1.364   0.047  -1.000
  1.85   1.82    655    528    1.24  22.65       2.8     0.2    1.860    2.583    1.786   0.019   0.000
  1.82   1.79    194    173    1.12   7.49       3.4     0.1   -2.792   -3.948   -2.792   0.072   0.000
 72.23   1.79 388046  36823   10.54  78.99      18.6     9.1    0.296    0.311    0.086   0.583*   0.003


Resolution limit suggested from cc1/2 fit (limit cc1/2=0.3): 2.18

               ----------Summary of merging statistics----------               

                                            Suggested   Low    High  Overall
High resolution limit                           2.18    5.92    2.18    1.79
Low resolution limit                           72.23   72.27    2.22   72.23
Completeness                                   94.5    90.1    87.3    79.0
Multiplicity                                   14.1    16.5     7.5    10.5
I/sigma                                        13.4    41.1     1.9     9.1
Rmerge(I)                                     0.260   0.107   1.452   0.296
Rmerge(I+/-)                                  0.260   0.106   1.450   0.293
Rmeas(I)                                      0.268   0.109   1.555   0.311
Rmeas(I+/-)                                   0.274   0.111   1.620   0.316
Rpim(I)                                       0.063   0.023   0.537   0.086
Rpim(I+/-)                                    0.084   0.032   0.699   0.108
CC half                                       0.899   0.998   0.311   0.583
Anomalous completeness                         62.8    74.4    48.9    44.4
Anomalous multiplicity                          8.4     9.0     4.8     6.7
Anomalous correlation                         0.000   0.008  -0.003   0.003
Anomalous slope                               0.791                   0.688
dF/F                                          0.100                   0.235
dI/s(dI)                                      0.683                   0.492
Total observations                           340548   19123    8309  388046
Total unique                                  24236    1161    1103   36823

Writing html report to dials.scale.html
Saving the scaled experiments to scaled.expt
Saving the scaled reflections to scaled.refl
See dials.github.io/dials_scale_user_guide.html for more info on scaling options
```

**Comment by graeme-winter** [comment]
Just run this

```
                                             Overall    Low     High
High resolution limit                           1.08    2.94    1.08
Low resolution limit                          102.23  102.69    1.10
Completeness                                   88.6   100.0    16.7
Multiplicity                                   20.2    25.0     1.1
I/sigma                                        21.9    63.6     1.7
Rmerge(I)                                     0.088   0.052   0.240
Rmerge(I+/-)                                  0.086   0.052   0.259
Rmeas(I)                                      0.090   0.053   0.333
Rmeas(I+/-)                                   0.090   0.053   0.363
Rpim(I)                                       0.018   0.011   0.229
Rpim(I+/-)                                    0.025   0.014   0.254
CC half                                       0.999   0.999   0.880
Anomalous completeness                         84.7   100.0     1.5
Anomalous multiplicity                         10.8    14.2     1.0
Anomalous correlation                        -0.037  -0.067   0.000
Anomalous slope                               0.956
dF/F                                          0.050
dI/s(dI)                                      0.799
Total observations                          1828128  138425     958
Total unique                                  90451    5533     842
```

So worked on 1st cut 🙂

**Comment by graeme-winter** [comment]
Change set looks sensible.

I wonder if we should pull the equivalent code out of xia2 & just use this?

Will go look at the change sets in more detail now.

File: algorithms/merging/merge.py [hunk 1]
```diff
@@ -17,7 +17,10 @@
 )
 from dials.util.filter_reflections import filter_reflection_table
 from dials.util.export_mtz import MADMergedMTZWriter, MergedMTZWriter
-from dials.report.analysis import make_merging_statistics_summary
+from dials.report.analysis import (
+    make_merging_statistics_summary,
+    make_xia2_style_statistics_summary,
```
Comment (graeme-winter) [inline algorithms/merging/merge.py:22]: Let's not call it xia2 style? Maybe `make_table_1_summary` or similar?

File: algorithms/merging/merge.py [hunk 2]
```diff
@@ -254,5 +257,6 @@ def merge_and_truncate(params, experiments, reflections):
                 logger.info(make_merging_statistics_summary(anom_stats))
             else:
                 logger.info(make_merging_statistics_summary(stats))
+            logger.info(make_xia2_style_statistics_summary(stats, anom_stats))
```
Comment (graeme-winter) [inline algorithms/merging/merge.py:260]: You tested this with centric data?

File: algorithms/scaling/observers.py [hunk 3]
```diff
@@ -42,8 +43,13 @@
     make_image_range_table,
 )
 from dials.algorithms.scaling.scale_and_filter import make_scaling_filtering_plots
+from dials.algorithms.scaling.scaling_library import (
+    merging_stats_from_scaled_array,
+    DialsMergingStatisticsError,
```
Comment (graeme-winter) [inline algorithms/scaling/observers.py:48]: Discussion topic: are custom specific exceptions a useful thing?

File: algorithms/scaling/observers.py [hunk 4]
```diff
@@ -191,9 +197,37 @@ def print_scaling_summary(self, scaling_script):
...
         )
         if MergingStatisticsObserver().data:
+            merging_stats = MergingStatisticsObserver().data["statistics"]
+            logger.info(make_merging_statistics_summary(merging_stats))
+            vals = cc_half_fit(merging_stats, limit=0.3)
+            r_cc = vals[0]
+            max_current_res = merging_stats.bins[-1].d_min
+            cut_merging_statistics_result = None
+            cut_anom_merging_statistics_result = None
+            if r_cc - max_current_res > 0.005:
```
Comment (graeme-winter) [inline algorithms/scaling/observers.py:207]: `0.005` because?

Comment (graeme-winter) [reply algorithms/scaling/observers.py:207]: Additional: I dislike magic numbers, and also the significance of these depends very much on the value under comparison

File: algorithms/scaling/observers.py [hunk 5]
```diff
@@ -191,9 +197,37 @@ def print_scaling_summary(self, scaling_script):
...
+                logger.info(
+                    "Resolution limit suggested from cc1/2 fit (limit cc1/2=0.3): %.2f",
```
Comment (graeme-winter) [inline algorithms/scaling/observers.py:209]: Want to use CC½ here?

File: algorithms/scaling/observers.py [hunk 6]
```diff
@@ -191,9 +197,37 @@ def print_scaling_summary(self, scaling_script):
...
+                    (
+                        cut_merging_statistics_result,
+                        cut_anom_merging_statistics_result,
+                    ) = merging_stats_from_scaled_array(
+                        scaling_script.scaled_miller_array.resolution_filter(
+                            d_min=r_cc
+                        ),
+                        scaling_script.params.output.merging.nbins,
+                        scaling_script.params.output.use_internal_variance,
+                    )
```
Comment (graeme-winter) [inline algorithms/scaling/observers.py:222]: Surely this screams at some kind of need for a named tuple or something? The line / variable lengths are absurd here 🙂

File: algorithms/scaling/observers.py [hunk 7]
```diff
@@ -243,7 +277,7 @@ def make_scaling_html(self, scaling_script):
             with open(html_file, "wb") as f:
                 f.write(html.encode("utf-8", "xmlcharrefreplace"))
         if json_file:
-            logger.info("Writing html report data to: %s", json_file)
+            logger.info("Writing html report data to %s", json_file)
```
Comment (graeme-winter) [inline algorithms/scaling/observers.py:280]: `html -> json_file` - surely some mistake? 🤔

File: newsfragments/1312.misc [hunk 8]
```diff
@@ -0,0 +1 @@
+Use xia2 reporting style for dials.scale and dials.merge merging statistics summary
```
Comment (graeme-winter) [inline newsfragments/1312.misc:1]: Prefer "add Table 1, following the style of output from `xia2`" or something (i.e. `xia2` does not own the style

File: report/analysis.py [hunk 9]
```diff
@@ -182,20 +183,169 @@ def _batch_bins_and_data(batches, values, function_to_apply):
     return batch_bins, data
 
 
+formats = collections.OrderedDict(
+    [
+        ("High resolution limit", " %7.2f"),
```
Comment (graeme-winter) [inline report/analysis.py:188]: Humbly ask if this static format is appropriate for very high resolution small molecule data?

File: report/analysis.py [hunk 10]
```diff
@@ -182,20 +183,169 @@ def _batch_bins_and_data(batches, values, function_to_apply):
...
+                    columns == 4 and len(row_data) == 2
+                ):  # place value in overall & suggest column
+                    row_data = [row_data[1]] + ([None] * (columns - 2)) + [row_data[0]]
+                row_format = [format_str] + [format_str.strip()] * (len(row_data) - 1)
+                formatted = " ".join(
+                    (f % k) if k is not None else (" " * len(f % 0))
+                    for f, k in zip(row_format, row_data)
+                )
+            except TypeError:
+                formatted = "(error)"
```
Comment (graeme-winter) [inline report/analysis.py:243]: 🤔  not sure I like this one...

File: report/analysis.py [hunk 11]
```diff
@@ -182,20 +183,169 @@ def _batch_bins_and_data(batches, values, function_to_apply):
...
+            result += k.ljust(44) + formatted + "\n"
+
+    return result
+
+
+def make_xia2_style_statistics_summary(
```
Comment (graeme-winter) [inline report/analysis.py:249]: ~xia2~

File: report/analysis.py [hunk 12]
```diff
@@ -182,20 +183,169 @@ def _batch_bins_and_data(batches, values, function_to_apply):
...
+    merging_statistics,
+    anomalous_statistics=None,
+    selected_statistics=None,
+    selected_anomalous_statistics=None,
+):
+
+    out = StringIO()
```
Comment (graeme-winter) [inline report/analysis.py:256]: Is doing this by means of a `StringIO` the most easy to understand solution?

File: report/test_analysis.py [hunk 13]
```diff
@@ -12,6 +13,11 @@
     batch_dependent_properties,
     reflection_tables_to_batch_dependent_properties,
     combined_table_to_batch_dependent_properties,
+    make_xia2_style_statistics_summary,
```
Comment (graeme-winter) [inline report/test_analysis.py:16]: You're screaming "long function names" at me here - could encapsulate them somehow? I appreciate that some of these predate the PR but you're taking over half a VT100 with some of those function names 😱

File: util/resolutionizer.py [hunk 14]
```diff
@@ -304,6 +304,73 @@ def miller_array_from_mtz(unmerged_mtz, params):
...
+            i = 0
+        else:
+            i += 1
+
+    if cc_half_fit == "tanh":
+        cc_f = tanh_fit(s_s[i:], cc_s[i:], iqr_multiplier=4)
+    else:
+        cc_f = fit(s_s[i:], cc_s[i:], 6)
+
+    logger.debug("rch: fits")
```
Comment (graeme-winter) [inline util/resolutionizer.py:355]: `rch:` ?

File: algorithms/scaling/observers.py (hunk 3 above)
Comment (jbeilstenedmands) [reply algorithms/scaling/observers.py:48]: Useful in this context I believe, as merging statistics code can throw a number of different exceptions which can be difficult to predict.

File: algorithms/scaling/observers.py (hunk 4 above)
Comment (jbeilstenedmands) [reply algorithms/scaling/observers.py:207]: We report resolution stats to 2dp, so this is meant to be "if the same to within two decimal places". xia2's magic number here is 0.004 🤷‍♂️ 🙃

File: algorithms/scaling/observers.py (hunk 7 above)
Comment (jbeilstenedmands) [reply algorithms/scaling/observers.py:280]: Actually this one is correct, the html report data can be saved to a json if requested.

File: algorithms/merging/merge.py (hunk 2 above)
Comment (jbeilstenedmands) [reply algorithms/merging/merge.py]: I hadn't, thanks, tested and fixed now

File: algorithms/merging/merge.py [hunk 15]
```diff
@@ -19,7 +19,7 @@
 from dials.util.export_mtz import MADMergedMTZWriter, MergedMTZWriter
 from dials.report.analysis import (
     make_merging_statistics_summary,
-    make_xia2_style_statistics_summary,
+    table_1_summary,
```
Comment (graeme-winter) [inline algorithms/merging/merge.py:22]: 👍

**Comment by jbeilstenedmands** [comment]
Fixed most issues now and marked as resolved. Anything left are things carried over from existing code and I think are fine to stay as is.

File: util/resolutionizer.py [hunk 16]
```diff
@@ -304,6 +304,73 @@ def miller_array_from_mtz(unmerged_mtz, params):
...
+    merging_statistics,
+    significance_level=0.1,
+    cc_half_method="half_dataset",
+    cc_half_fit="tanh",
+    limit=0.3,
+):
+
+    if cc_half_method == "sigma_tau":
+        cc_s = flex.double(
+            [b.cc_one_half_sigma_tau for b in merging_statistics.bins]
```
Comment (Anthchirp) [inline util/resolutionizer.py:317]:
```suggestion
            b.cc_one_half_sigma_tau for b in merging_statistics.bins
```
flex constructors work with generators, no need to explicitly construct a list and then throw it away. Applies to this line and a couple more times down the file.

File: report/analysis.py [hunk 17]
```diff
@@ -182,28 +181,184 @@ def _batch_bins_and_data(batches, values, function_to_apply):
     return batch_bins, data
 
 
-def make_merging_statistics_summary(dataset_statistics):
-    """Format merging statistics information into an output string."""
+formats = collections.OrderedDict(
```
Comment (Anthchirp) [inline report/analysis.py:184]: CPython 3.6+ (and Python 3.7+) dictionaries are ordered by default. No more need to use `OrderedDict`.

**Pull request merged**